### PR TITLE
Implementing gradient descent for the ephemeral constants

### PR DESCRIPTION
**Pull request by darioizzo**

### CODE REVIEW

File: include/dcgp/problems/symbolic_regression.hpp [hunk 1]
```diff
@@ -75,33 +112,98 @@ class symbolic_regression
...
+            eph_val.emplace_back(x[i], m_dcgp.get_eph_symb()[i], 1u); // Only first derivative is needed
+        }
+        m_dcgp.set_eph_val(eph_val);
         // 3 - We compute the MSE loss splitting the data in n batches.
         // TODO: make this work also when m_parallel_batches does not divide exactly the data size.
-        retval[0] = m_cgp.loss(m_points, m_labels, "MSE", m_parallel_batches);
+        auto loss = m_dcgp.loss(m_dpoints, m_dlabels, "MSE", m_parallel_batches);
+        // since we have also computed the loss value, we store it in a cache so that fitness can be called and
+        // not cause reavaluation of a cgp.
+        m_cache = decltype(m_cache){x, pagmo::vector_double{loss.constant_cf()}};
```
Comment (bluescarni) [inline include/dcgp/problems/symbolic_regression.hpp]: Here it's better to assign directly the members of the pair rather than creating a new pair and then move-assign it.

```c++
m_cache.first = x;
m_cache.second = pagmo::vector_double{loss.constant_cf()};
```

Even better would be to avoid altogether the creation of a temporary ``vector_double``. Perhaps something like

```c++
m_cache.second.resize(1);
m_cache.second[0] = loss.constant_cf();
```

Comment (darioizzo) [reply include/dcgp/problems/symbolic_regression.hpp]: yeps, implemented thanks!
